File: leetcode/2001-3000/2001-2100/2017-grid-game/GridGame.java
public class GridGame {
  public static void main(String[] args) {
    int[][] grid = {
        { 1, 3, 1, 15 },
        { 1, 3, 3, 1 }
    };

    System.out.println(new Solution().gridGame(grid));
  }
}

class Solution {
  public long gridGame(int[][] grid) {
    int length = grid[0].length;
    long max = Long.MAX_VALUE;

    long right = 0;
    long left = 0;

    for (int i = 0; i < length; i++) {
      right += grid[0][i];
    }

    /* i is the transition point of the first robot. */
    for (int i = 0; i < length; i++) {
      right -= grid[0][i];
      max = Math.min(max, Math.max(left, right));
      left += grid[1][i];
    }

    return max;
  }
}

File: leetcode/2001-3000/2001-2100/2017-grid-game/GridGame.ts
function gridGame(grid: number[][]): number {
  let left = 0;
  let right = grid[0].reduce((accumulated, value) => accumulated + value, 0);
  let max = Number.MAX_VALUE;

  for (let i = 0; i < grid[0].length; i++) {
    right -= grid[0][i];
    max = Math.min(max, Math.max(left, right));
    left += grid[1][i];
  }

  return max;
};

console.log(gridGame([[1, 3, 1, 15], [1, 3, 3, 1]]))
